docs/source/java/substrait.rst

```diff
@@ -102,6 +104,349 @@ Here is an example of a Java program that queries a Parquet file using Java Subs
+    import org.apache.arrow.memory.BufferAllocator;
+    import org.apache.arrow.memory.RootAllocator;
+    import org.apache.arrow.vector.ipc.ArrowReader;
+
+    import java.nio.ByteBuffer;
+    import java.util.ArrayList;
+    import java.util.Arrays;
+    import java.util.Base64;
+    import java.util.HashMap;
+    import java.util.List;
+    import java.util.Optional;
+
+    public class ClientSubstraitExtendedExpressionsCookbook {
+        public static void main(String[] args) throws Exception {
+            // project and filter dataset using extended expression definition - 03 Expressions:
+            // Expression 01 - CONCAT: N_NAME || ' - ' || N_COMMENT = col 1 || ' - ' || col 3
+            // Expression 02 - ADD: N_REGIONKEY + 10 = col 1 + 10
+            // Expression 03 - FILTER: N_NATIONKEY > 18 = col 3 > 18
+            projectAndFilterDataset();
+        }
+
+        public static void projectAndFilterDataset() {
+            String uri = "file:///Users/data/tpch_parquet/nation.parquet";
+            ScanOptions options = new ScanOptions.Builder(/*batchSize*/ 32768)
+                    .columns(Optional.empty())
+                    .substraitFilter(getSubstraitExpressionFilter())
+                    .substraitProjection(getSubstraitExpressionProjection())
+                    .build();
+            try (
+                    BufferAllocator allocator = new RootAllocator();
+                    DatasetFactory datasetFactory = new FileSystemDatasetFactory(
+                            allocator, NativeMemoryPool.getDefault(),
+                            FileFormat.PARQUET, uri);
+                    Dataset dataset = datasetFactory.finish();
+                    Scanner scanner = dataset.newScan(options);
+                    ArrowReader reader = scanner.scanBatches()
+            ) {
+                while (reader.loadNextBatch()) {
+                    System.out.println(
+                            reader.getVectorSchemaRoot().contentToTSVString());
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+
+        private static ByteBuffer getSubstraitExpressionProjection() {
+            // Expression: N_REGIONKEY + 10 = col 3 + 10
+            Expression.Builder selectionBuilderProjectOne = Expression.newBuilder().
+                    setSelection(
+                            Expression.FieldReference.newBuilder().
+                                    setDirectReference(
+                                            Expression.ReferenceSegment.newBuilder().
+                                                    setStructField(
+                                                            Expression.ReferenceSegment.StructField.newBuilder().setField(
+                                                                    2)
+                                                    )
+                                    )
+                    );
```

Comment:
nit: how are you formatting examples? I think it would make sense to just use `google-java-format` which has a more compact style. In our docs, going too far to the right is unreadable.

Reply:
changed